### PR TITLE
update vendor spf13/cobra to enforce required flags

### LINKED ISSUE #121: Autocomplete issues with --from-file and --from-literal
**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.2", GitCommit:"bdaeafa71f6c7c04636251031f93464384d54963", GitTreeState:"clean", BuildDate:"2017-10-24T21:07:53Z", GoVersion:"go1.9.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.0", GitCommit:"0b9efaeb34a2fc51ff8e4d34ad9bc6375459c4a4", GitTreeState:"dirty", BuildDate:"2017-10-17T15:09:55Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: minikube
- **OS** (e.g. from /etc/os-release): macOS Sierra, 10.12.6
- **Kernel** (e.g. `uname -a`): `Darwin WDFM31654402A 16.7.0 Darwin Kernel Version 16.7.0: Wed Oct  4 00:17:00 PDT 2017; root:xnu-3789.71.6~1/RELEASE_X86_64 x86_64`
- **Install tools**: n/a
- **Others**: `GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin16)`

**What happened**:
I was following along the Kubernetes Up & Running book and found that autocomplete for some of the configmaps and secrets examples doesn't work as expected.

**What you expected to happen**:
I was expecting the autocomplete to work for the `--from-file` and `--from-literal` parameters, however, I got some bash errors.

**How to reproduce it** (as minimally and precisely as possible):
* Type:
```
kubectl create configmap my-config --from-file=<tab>
```
Result:
```
kubectl create configmap my-config --from-file=--from-file= (instead of a list of files in the local directory)
```
* Type:
```
kubectl create configmap my-config --from-file=my-config.txt --from-li<tab>
```
Result (my-config.txt should of course exist in the local directory):
```
kubectl create configmap my-config --from-file=my-config.txt --from-li-bash: --from-file=: syntax error: operand expected (error token is "=")
```

Similar behavior for `kubectl create secret`.

### LINKED ISSUE #54855: Deprecated usage message is not getting printed
**Is this a BUG REPORT or FEATURE REQUEST?**:

 /kind bug

**What happened**:
Deprecated usage message is not getting printed , such as --container-port in kubectl export, --interactive in kubectl log,  --experimental-use-openapi-print-columns in kubectl get.

Because cobra is used in kubectl, the issue is imported by cobra.
https://github.com/spf13/cobra/issues/463
https://github.com/spf13/cobra/pull/466

Should we update cobra in vendor?

$ kubectl expose --help
Expose a resource as a new Kubernetes service. 
...
Options:
      ...
      --container-port='': Synonym for --target-port

$ kubectl log --help
...
Options:
     ...
      --interactive=false: If true, prompt the user for input when required.

**What you expected to happen**:
$ kubectl expose --help
Expose a resource as a new Kubernetes service. 
...
Options:
      ...
      --container-port='':  --container-port will be removed in the future, please use --target-port instead

$ kubectl log --help
...
Options:
     ...
      --interactive=false: This flag is no longer respected and there is no replacement.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.2", GitCommit:"bdaeafa71f6c7c04636251031f93464384d54963", GitTreeState:"clean", BuildDate:"2017-10-24T19:48:57Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:

spf13/cobra#502 has enforced checking flags that marked as required, an error will be raised if unset.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:fixes #54855
xref #48400
fixes kubernetes/kubectl#121 

**Special notes for your reviewer**:
/assign @liggitt @eparis 

**Release note**:

```release-note
kubectl now enforces required flags at a more fundamental level
```

### CODE REVIEW

**Comment by eparis** [comment]
/ok-to-test
I thought we also needed to bump pflag...

**Comment by eparis** [comment]
CI says I'm right and we need to bump pflag as well.

**Comment by dixudx** [comment]
@eparis `func (c *Command) Name()` has been changed in spf13/cobra#422. This is causing [deprecated alias](https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/kubectl/cmd/cmd_test.go#L672-L720) not working as expected. Please refer to [my comment]Ihttps://github.com/spf13/cobra/pull/422#discussion_r143918343)

**Comment by dixudx** [comment]
@eparis @liggitt PTAL. It's ready for review. Thanks

**Comment by dixudx** [comment]
@eparis @liggitt PTAL. It's ready for review. Thanks

**Comment by dixudx** [comment]
ping @liggitt @smarterclayton @wojtek-t PTAL. Thanks

**Comment by dixudx** [comment]
ping @liggitt @smarterclayton @kubernetes/sig-cli-maintainers  PTAL. Thanks.

**Comment by yuexiao-wang** [comment]
@dixudx Could you tell me how to update  spf13/cobra ? Because I have some problem to update it, pleae show the detail steps.  Thanks

**Comment by dixudx** [comment]
@yuexiao-wang The bugs you mentioned in #54855 has been fixed in this PR. So there is no need to submit another PR.

> how to update spf13/cobra

Kubernetes uses [godep](https://github.com/tools/godep) to manage all the deps. Here is the steps you can follow if you have other vendor/deps to update.

```sh
# step 1, restore all deps
$ ./hack/godep-restore.sh
# step2, update your dep in your GOPATH
$ <update local dep>
# step3, save the deps
$ ./hack/godep-save.sh
```

**Comment by yuexiao-wang** [comment]
Got it and thanks

**Comment by dixudx** [comment]
ping @eparis @liggitt 

Rebased. PTAL. Thanks.

**Comment by shiywang** [comment]
@dixudx could you also change your description also to fix this kubernetes/kubectl#121  btw ?

**Comment by shiywang** [comment]
@dixudx also could you also sig-cli next time, so we could avoid duplicate works ? thanks

**Comment by fabianofranz** [comment]
This introduces an issue with the usage string:

```
kubectl describe (-f FILENAME | TYPE [NAME_PREFIX | -l label] | TYPE/NAME) [flags] [options]
```

Note both `[flags]` and `[options]` where we just want `[options]`. Apparently something changed in Cobra templates, or something like that.

**Comment by dixudx** [comment]
> Note both [flags] and [options] where we just want [options]. Apparently something changed in Cobra templates, or something like that.

@fabianofranz Yes, you're right. spf13/cobra#422 introduced this change with adding `[flags]` in UseLine. That's why the usage string gets changed.

 But I don't think this change will break `kubectl`.

**Comment by shiywang** [comment]
I agree with @fabianofranz, although this won't break the functionality, but we don't want both `[flags] [options]` exist in every command, looks like a bug to user.

**Comment by fabianofranz** [comment]
> looks like a bug to user

Correct.

**Comment by dixudx** [comment]
> Note both [flags] and [options] where we just want [options]. Apparently something changed in Cobra templates, or something like that.
> looks like a bug to user

spf13/cobra#552 adds a compromise way to disable it. 

Already updated it. PTAL @shiywang @fabianofranz

**Comment by dixudx** [comment]
@kubernetes/sig-cli-maintainers PTAL. Thanks

File: pkg/kubectl/cmd/rollingupdate.go
Comment (shiywang) [inline pkg/kubectl/cmd/rollingupdate.go]: why remove this ?

File: docs/.generated_docs
Comment (shiywang) [inline docs/.generated_docs]: @kubernetes/sig-cli-maintainers is it ok to remove those ? although `run-container` is a deprecated command [here](https://github.com/kubernetes/kubernetes/blob/81631eceb80aace00dfb221b13440c7f658a18c7/pkg/kubectl/cmd/cmd.go#L250)

File: docs/man/man1/kubectl-run-container.1
Comment (shiywang) [inline docs/man/man1/kubectl-run-container.1]: ok, I see, could you squash this into commit `fix wrong required flags`  ?

File: cmd/genkubedocs/postprocessing.go
Comment (shiywang) [inline cmd/genkubedocs/postprocessing.go]: cobra changed ?  I think we could have three commit, 1.pflag gen, 2.cobra gen, 3.the code you changed

Comment (dixudx) [reply cmd/genkubedocs/postprocessing.go]: Yes, method rename. Please see my [last commit](https://github.com/kubernetes/kubernetes/pull/53631/commits/f31ec301c36ffac215242fb9bf44b9e0633d9a8c).

File: pkg/kubectl/cmd/rollingupdate.go
Comment (dixudx) [reply pkg/kubectl/cmd/rollingupdate.go]: Because it is not a required flag. You can also issue the command with `kubectl rolling-update frontend-v1 -f xxx`.

Please refer to the template example in `rollingupdate.go`.

**Comment by shiywang** [comment]
/lgtm
@dixudx  thank you for this great job !!

**Comment by dixudx** [comment]
ping @wojtek-t @smarterclayton for approval. Thanks.

**Comment by fabianofranz** [comment]
@dixudx Maybe add a test that checks for that duplicated `[flags] [options]` issue?

**Comment by dixudx** [comment]
@fabianofranz make sense. And this will help use make sure the help message is what k8s expected.

**Comment by dixudx** [comment]
@fabianofranz Test added. PTAL.

/cc @kubernetes/sig-cli-maintainers

File: pkg/kubectl/cmd/describe_test.go
Comment (shiywang) [inline pkg/kubectl/cmd/describe_test.go]: why not `describe (-f FILENAME | TYPE [NAME_PREFIX | -l label] | TYPE/NAME) [options]` ?

Comment (dixudx) [reply pkg/kubectl/cmd/describe_test.go]: Good catch. Thanks.

Comment (dixudx) [reply pkg/kubectl/cmd/describe_test.go]: @shiywang `[options]` is added by kubectl template automatically, not belonging to `cobra`.

https://github.com/kubernetes/kubernetes/blob/c933067cd3cfa81349446b929816f732aa145b90/pkg/kubectl/cmd/templates/templater.go#L198-L205

This test will only cover testing `[flags]` in `useline`. No need to change to `describe (-f FILENAME | TYPE [NAME_PREFIX | -l label] | TYPE/NAME) [options]`.

Comment (shiywang) [reply pkg/kubectl/cmd/describe_test.go]: oh, sorry, my mistake, looks good

Comment (shiywang) [reply pkg/kubectl/cmd/describe_test.go]: wait, I still don't think this will catch the error, did a little test on my laptop
```
var t *testing.T
got := "describe (-f FILENAME | TYPE [NAME_PREFIX | -l label] | TYPE/NAME) [flags]"
expected := "describe (-f FILENAME | TYPE [NAME_PREFIX | -l label] | TYPE/NAME)"
if !strings.Contains(got, expected) {
	t.Errorf("Expected to contain: \n %v\nGot:\n %v\n", expected, got)
}
```

Comment (dixudx) [reply pkg/kubectl/cmd/describe_test.go]: > maybe we only need to cover it doesn't contains [flags]

@shiywang make sense.

**Comment by dixudx** [comment]
ping @shiywang @fabianofranz. Please have a review. Thanks.

/cc @kubernetes/sig-cli-maintainers

**Comment by dixudx** [comment]
/cc @kubernetes/sig-cli-maintainers
ping @liggitt @fabianofranz PTAL. Thanks.

**Comment by dixudx** [comment]
ping @smarterclayton @wojtek-t for approval. Thanks.

**Comment by dixudx** [comment]
ping @smarterclayton @wojtek-t @brendandburns @thockin for approval. Thanks.

**Comment by dixudx** [comment]
ping @smarterclayton @wojtek-t @brendandburns @thockin for approval. Thanks.

**Comment by thockin** [comment]
I can approve once it is LGTM - the godeps changes are OK

**Comment by dixudx** [comment]
ping @thockin for approval. Thanks.

**Comment by dixudx** [comment]
ping @smarterclayton @wojtek-t @thockin for approval. Thanks.

**Comment by dixudx** [comment]
ping @smarterclayton @dchen1107 for approval. Thanks.

**Comment by jbeda** [comment]
I'll do the approval here if others aren't around.  Looks totally sane.

I'd like to see the release note updated to something that users will actually be able to understand.  Something like "kubectl now enforces required flags at a more fundamental level".

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by soltysh** [comment]
@jbeda || @thockin it's all your for approval

**Pull request merged**
